### PR TITLE
[SPARK-47147][PYTHON][SQL] Fix PySpark collated string conversion error

### PR DESCRIPTION
**Pull request by nikolamand-db**
### What changes were proposed in this pull request?

When running Pyspark shell in non-Spark Connect mode, query `SELECT 'abc' COLLATE 'UCS_BASIC_LCASE'` produces the following error:
```
AssertionError: Undefined error message parameter for error class: CANNOT_PARSE_DATATYPE. Parameters: {'error': "Undefined error message parameter for error class: CANNOT_PARSE_DATATYPE. Parameters: {'error': 'string(UCS_BASIC_LCASE)'}"}
```
Fix the error by updating `StringType` class to include collation id support. Additional changes ensure conversions work properly.

### Why are the changes needed?

To fix collations-related error in PySpark shell.

### Does this PR introduce _any_ user-facing change?

Yes, the described sql query no longer throws an exception.

### How was this patch tested?

Manually tested with PySpark shell, additional assertions are added to type tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -244,10 +244,39 @@ class FractionalType(NumericType):
...
+        return "(" + StringType.collationNames[collationId] + ")" if collationId != 0 else ""
+
+    def collationNameToId(collationName: str) -> int:
+        return StringType.collationNames.index(collationName) if collationName != "" else 0
+
+    def simpleString(self) -> str:
+        return "string%s" % (self.collationIdToName(self.collationId))
+
+    def jsonValue(self) -> str:
+        return "string%s" % (self.collationIdToName(self.collationId))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:276]: Is this matched with `DataType.COLLATED_STRING_TYPE` in Scala? It has to be same and able to parse via `DataType.parseDataType`

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -244,10 +244,39 @@ class FractionalType(NumericType):
...
+        self.collationId = collationId
+
+    def collationIdToName(self, collationId: int) -> str:
+        return "(" + StringType.collationNames[collationId] + ")" if collationId != 0 else ""
+
+    def collationNameToId(collationName: str) -> int:
+        return StringType.collationNames.index(collationName) if collationName != "" else 0
+
+    def simpleString(self) -> str:
+        return "string%s" % (self.collationIdToName(self.collationId))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:273]: To confirm, is this able to parse this string via our SQL parser?

File: python/pyspark/sql/tests/test_types.py [hunk 3]
```diff
@@ -849,6 +849,7 @@ def test_parse_datatype_string(self):
             if k != "varchar" and k != "char":
                 self.assertEqual(t(), _parse_datatype_string(k))
         self.assertEqual(IntegerType(), _parse_datatype_string("int"))
+        self.assertEqual(StringType(0), _parse_datatype_string("string"))
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_types.py:852]: Can we add other cases like `string(0)`?

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (nikolamand-db) [reply python/pyspark/sql/types.py]: @HyukjinKwon Aligned this with collated string regex from Scala and added more tests to confirm parsing works as you suggested, please check the latest changes.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**